File: chapter4/sorular.pl
% Örnek listeler
liste1([5,3,8,12,2,7,6]).
liste2([4,-6,3,7,-11,25,-45,-20]).
liste3([1, 1, 2, 2, 2, 1, 1, 3]).
% Listenin ilk elemanını veren yüklem
ilkeleman([B|K],B).
% Listenin son elemanını veren yüklem
soneleman([S],S) :- !.
soneleman([B|K],Son) :- soneleman(K,Son).
% Listenin sondan bir önceki elemanını veren yüklem
sondanbironce([O1,O0], O1) :- !.
sondanbironce([B|K],Son) :- sondanbironce(K, Son).
% Listedeki en küçük sayıyı veren yüklem
% enkucuk(+Liste:list, -EnKucuk:int)
enkucuk([B],B) :- !.
enkucuk([B|K],EK) :- enkucuk(K, KEK), EK is min(B,KEK).

% min(+A:int,+B:int,-C:int)
min(A,B,A) :- A<B, !.
min(A,B,B) :- A>=B.

enkucuk2([B], B) :- !.
enkucuk2([B|K], EK) :- enkucuk2(K, KEK), min(B, KEK, EK).

% Listedeki elemanların toplamını veren yüklem
% toplam(+Liste:list, -Toplam:int)
toplam([], 0).
toplam([B|K], Toplam) :- toplam(K, KuyrukToplam), Toplam is B+KuyrukToplam.
% Listedeki elemanların karesini veren yüklem
% kare(+Liste:list, -Liste2:list)
kare([], []).
kare([B|K], [B2|K2]) :- kare(K, K2), B2 is B*B.
% Listedeki negatif sayıları filtreleyen yüklem
% negatif(+Liste:list, -Liste2:list)
negatif([],[]).
negatif([B|K],K2) :- B<0, negatif(K, K2), !.
negatif([B|K],[B|K2]) :- B>=0, negatif(K,K2).
% Listede ardışık tekrar eden elemanları eleyen/ayıklayan yüklem
%    ele([1, 1, 2, 2, 2, 1, 1, 3], [1, 2, 1, 3]).
ele([],[]).
ele([T],[T]).
ele([B,B|K],K2) :- ele([B|K], K2), !.
ele([A,B|K],[A|K2]) :- A \= B, ele([B|K], K2).

File: chapter3/sorular.hs
liste = [1..10]
liste2 = [5,2,8,6,5,4,12,3]
-- 1 ile parametre olarak verilen n sayısı arasındaki tek sayıların toplamını hesaplayan fonksiyon
topla 1 = 1
topla n = if odd n then n+topla (n-2) else topla (n-1)
--liste üreteci ile yapımı
topla2 n = sum [x | x<-[1..n], odd x]
-- Parametre olarak verilen listedeki elemanların toplamını döndüren fonksiyon
listetopla [] = 0
listetopla (b:k) = b+listetopla k
-- Parametre olarak verilen listenin en küçük elemanını veren fonksiyon
enkucuk [x] = x
enkucuk (b:k) = min b (enkucuk k)
-- Parametre olarak verilen listedeki tek sayıların bulunduğu listeyi döndüren fonksiyon
teksayilar liste = [x | x <- liste, odd x]

teksayilar2 [] = []
teksayilar2 (b:k) = if odd b then b:teksayilar2 k else teksayilar2 k

teksayilar3 [] = []
teksayilar3 (b:k)
    | odd b = b:teksayilar3 k
    | otherwise = teksayilar3 k
-- Bir fonksiyon ve bir listeyi parametre olarak alan ve listeyi fonksiyona göre filtreleyen fonksiyon
filtrele _ [] = []
filtrele f (b:k)
    | f b = b:filtrele f k
    | otherwise = filtrele f k
-- map fonksiyonunu kullanan ve parametre olarak verilen listedeki sayıların karesini alıp liste olarak döndüren fonksiyon
kare liste = map (\x -> x*x) liste

-- İkili arama ağacı kullanıcı tanımlı veri türü
data BST a = Nil | Node a (BST a) (BST a) deriving (Show)
agac1 = Node 12 (Node 5 (Node 3 Nil Nil) (Node 8 Nil Nil)) (Node 25 (Node 14 Nil Nil) Nil)
--        12
--       /   \
--      5    25
--     / \   /
--    3   8 14 
-- İkili arama ağacında arama yapan fonksiyon
--ara:: BST a -> Bool
ara Nil _ = False
ara (Node deger sol sag) aranacak
    | deger == aranacak = True
    | aranacak < deger = ara sol aranacak
    | otherwise = ara sag aranacak
-- İkili arama ağacına ekleme yapan fonksiyon
ekle Nil eklenecek = Node eklenecek Nil Nil
ekle (Node deger sol sag) eklenecek
    | deger == eklenecek = Node deger sol sag --error "Değer zaten ağaçta mevcut"
    | eklenecek < deger = Node deger (ekle sol eklenecek) sag
    | otherwise = Node deger sol (ekle sag eklenecek)
-- İkili arama ağacından değer silen fonksiyon
sil